README.md
# [Perfil Parlamentar](https://perfil.parlametria.org/)

[![Build Status](https://travis-ci.org/analytics-ufcg/voz-ativa.svg?branch=master)](https://travis-ci.org/analytics-ufcg/voz-ativa)

[Documentação da API](https://perfil.parlametria.org/api/swagger/docs/)

# Sobre o projeto

O Perfil Parlamentar exibe informações sobre as pessoas que compõem a Câmara dos Deputados e o Senado do congresso brasileiro.

# Desenvolvimento

O projeto foi desenvolvido utilizando a arquitetura PEAN (Postgres Express Angular e Node).

Instale o [docker](https://docs.docker.com/install/linux/docker-ce/ubuntu/#install-docker-ce) e o [docker-compose](https://docs.docker.com/compose/install/).

## Inicialização do banco de dados

No [repositório](https://github.com/analytics-ufcg/vozativa-dados) de dados do projeto siga as instruções do README para inicializar o banco de dados usando docker.

## Configuração das variáveis de ambiente

Faça uma cópia arquivo `variables.env.sample` para o arquivo `variables.env` e preencha com as chaves para as variáveis de ambiente. 

## Iniciando docker

Após a definição das variáveis e com o banco de dados executando via docker. Rode com o docker executando:

```
docker-compose up
```

Pronto, o desenvolvimento já pode ser iniciado!

A api estará disponível em: localhost:5000/api
O frontend estará disponível em: localhost:4200

Caso for preciso reconstruir as imagens (backend e frontend) basta fazer:

```
docker-compose up --build
```

**Limitação**: Alterações no arquivo package.json (do frontend ou do backend) exigem um rebuild (reconstrução) da imagem com o comando citado acima.

### Comandos úteis

Caso você queira parar os containers e remover os volumes execute:

```
docker-compose down --volumes
```

Para visualizar os containers rodando:

```
docker ps
```

Para executar comandos num shell dentro do container:

```
docker exec -it <container_id> sh
```

Para matar um container

```
docker kill <container_id>
```

# Desenvolvimento sem Docker

## Configuração das variáveis de ambiente

É necessário exportar as variáveis de ambiente para que o projeto possa ser executado. Essa é a lista de variáveis necessárias:

- SECRET_OR_KEY
- POSTGRESURI
- FACEBOOK_APP_ID
- FACEBOOK_APP_SECRET
- GOOGLE_APP_ID
- GOOGLE_APP_SECRET

Para exportar use, como exemplo, o comando abaixo (no terminal):

```
export SECRET_OR_KEY="umsupersegredo"
```

Obs: a variável POSTGRESURI deve ter o seguinte formato `postgres://<username>:<password>@postgres:5432/vozativa`

## Instalação de dependências

### Para o backend

```
npm install
```

### Para o frontend

```
npm run client-install
```

## Execução do projeto

### Para execução do backend

```
npm run server
```

### Para execução do frontend

```
npm run client
```

# Testes

Os testes são escritos usando [chai](https://www.chaijs.com) e rodam via [docker-compose](https://github.com/mochajs/mocha).

### Executar testes

```
npm test
```


CHANGELOG.md
# Changelog

## 3.5.1 (2020-06-23)

- [x] Adiciona endpoint para recuperar peso político para todos os parlamentares

## 3.5.0 (2020-06-01)

- [x] Atualiza Angular e demais pacotes para a versão 9.x

## 3.4.1 (2020-05-05)

- [x] Documentação da API

## 3.4.0 (2019-11-26)

- [x] Lançamento do Perfil Parlamentar: nova identidade e descrição do aplicativo
- [x] Adiciona trajetória política do parlamentar com filiações, mandatos e patrimônio
- [x] Altera a navegação geral do aplicativo
- [x] Referencia o Parlametria

## 3.3.0 (2019-11-13)

- [x] Adiciona investimento partidário dos parlamentares
- [x] Disponibiliza dados de sócios de empresas e doações via API
- [x] Resolução de bugs

## 3.2.2 (2019-10-21)

- [x] Correção de textos
- [x] Resolução de bugs

## 3.2.0 (2019-10-16)

- [x] Expande o aplicativo adicionando dados e visualizações do Senado

## 3.1.1 (2019-08-16)

- [x] Atualiza descrição do app
- [x] Resolução de bugs

## 3.1.0 (2019-08-08)

- [x] Cria filtro de lideranças partidárias
- [x] Cria filtro de cargos em comissões
- [x] Permite escolher ordenação da aderência (ascendente/descendente)
- [x] Atualiza 'Sobre'
- [x] Resolução de bugs

## 3.0.1 (2019-07-24)

- [x] Considera orientação null como voto liberado
- [x] Adiciona informações detalhadas sobre votações
- [x] Cria serviço de atualização da aplicação usando service worker
- [x] Exibe data da votação e link para proposição na Câmara

## 3.0.0 (2019-07-12)

- [x] Reformula aplicação: exibe informações sobre a aderência dos deputados ao governo
- [x] Redefine leiaute do filtro tema
- [x] Exibe informações dos votos dos deputados e do governo no perfil de cada parlamentar
- [x] Adiciona filtro de tema na seção de aderência do perfil do parlamentar

## 2.2.0 (2019-06-25)

- [x] Divide o perfil do parlamentar em subsessões
- [x] Adiciona taxa de aderência do parlamentar ao partido e ao Governo
- [x] Adiciona informações sobre lideranças partidárias

## 2.1.0 (2019-06-04)

- [x] Adiciona visualização do alinhamento com o congresso
- [x] Novo pacote de ícones :sparkles:

## 2.0.0 (2019-05-24)

- [x] Adiciona Google Analytics
- [x] Adiciona informação do cargo do parlamentar na comissão
- [x] Permite filtragem de parlamentares a partir da comissão selecionada no filtro
- [x] Adiciona atalhos de teclado ao questionário :abc:
- [x] Adiciona configuração do Travis CI
- [x] Configura e adiciona service worker à aplicação :construction_worker:
- [x] Otimiza carregamento de imagens dos parlamentares (lazy load)
- [x] Cria estrutura para testes
- [x] Verifica o ambiente para aplicar ou não SSL
- [x] Adiciona forma de usar docker para o ambiente de desenvolvimento do projeto
- [x] Instala Bootstrap 4.2.1
- [x] Cria estrutura para frontend em Angular 7
- [x] Novo layout :star2:

## 1.5.4 (2019-04-23)

- [x] Adiciona permissão de cors para o ambiente de desenvolvimento do leggo
- [x] Adiciona endpoint para retornar mapeamento entre id dos parlamentares

## [1.5.3](https://github.com/analytics-ufcg/voz-ativa/pull/224) (2019-03-29)

- [x] Adiciona informação de Comissões no perfil do Parlamentar
- [x] Filtra parlamentares por comissão
- [x] Atualiza 'Sobre'
- [x] Melhora o tempo de requisição das respostas dos parlamentares
- [x] Remove URLs de compartilhamento
- [x] Atualiza textos dos filtros


server.js
const express = require("express");
const http = require("http");
const bodyParser = require("body-parser");
const path = require("path");
const logger = require("heroku-logger");
const cors = require("cors");
const passport = require("passport");
const compression = require("compression");
const forceSsl = require('force-ssl-heroku');

const perguntas = require("./server/routes/api/perguntas");
const auth = require("./server/routes/api/auth");
const usuarios = require("./server/routes/api/usuarios");
const temas = require("./server/routes/api/temas");
const alinhamento = require("./server/routes/api/alinhamento");
const comissoes = require("./server/routes/api/comissoes");
const parlamentares = require("./server/routes/api/parlamentares");
const aderencia = require("./server/routes/api/aderencia");
const orientacoes = require("./server/routes/api/orientacoes");
const liderancas = require("./server/routes/api/liderancas");
const perfil = require("./server/routes/api/perfil");
const investimentoPartidario = require("./server/routes/api/investimentoPartidario");
const buscaParlamentar = require("./server/routes/api/busca-parlamentar");
const votacoes = require("./server/routes/api/votacoes");
const swagger = require("./server/routes/api/swagger");

const app = express();
app.use(forceSsl);

app.use(compression());
var db = require("./server/models/index");

const corsOptions = {
  origin: ['http://localhost:4200', 'http://localhost:3000', 'https://front.dev.leggo.org.br', 'https://leggo.org.br'],
  methods: "GET,HEAD,PUT,PATCH,POST,DELETE",
  credentials: true,
  allowedHeaders: ["Content-Type", "Authorization"],
  exposedHeaders: ["authorization", ]
};
app.use(cors(corsOptions));

// Body parser middleware
app.use(bodyParser.urlencoded({ extended: false }));
app.use(passport.initialize());
app.use(bodyParser.json());

// Testa conexão com o BD
db.sequelize
  .authenticate()
  .then(() => {
    logger.info("Conexão com BD estabelecida com sucesso.");
  })
  .catch(err => {
    logger.error("Não foi possível conectar com o BD: ", err);
  });

// Usar as rotas
app.use("/api/perguntas", perguntas);
app.use("/api/usuarios", usuarios);
app.use("/api/auth", auth);
app.use("/api/temas", temas);
app.use("/api/alinhamento", alinhamento);
app.use("/api/comissoes", comissoes);
app.use("/api/parlamentares", parlamentares);
app.use("/api/aderencia", aderencia);
app.use("/api/orientacoes", orientacoes);
app.use("/api/liderancas", liderancas);
app.use("/api/perfil", perfil);
app.use("/api/investimento", investimentoPartidario);
app.use("/api/busca-parlamentar", buscaParlamentar);
app.use("/api/votacoes", votacoes);
app.use("/api/swagger", swagger);


// Set static folder
app.use(express.static("client/build"));

app.get("*", (req, res) => {
  res.sendFile(path.resolve(__dirname, "client", "build", "index.html"));
});

const port = process.env.PORT || 5000;

app.listen(port, () => logger.info(`Servidor rodando na porta ${port}`));
